panther_hardware_interfaces/test/test_gpio_controller.cpp

```diff
@@ -0,0 +1,161 @@
+  gpio_controller_wrapper_ = std::make_unique<GPIOControllerWrapper>();
+
+  if (GetRobotVersion() < 1.2 - std::numeric_limits<float>::epsilon()) {
+    throw std::runtime_error("Tests for this robot versions are not implemented");
+  }
+
+  gpio_controller_wrapper_->Start();
+}
+
+float TestGPIOController::GetRobotVersion()
+{
+  const char * robot_version_env = std::getenv("PANTHER_ROBOT_VERSION");
+
+  if (!robot_version_env) {
+    throw std::runtime_error("Can't read 'PANTHER_ROBOT_VERSION' environment variable");
+  }
+
+  return std::stof(robot_version_env);
+}
+
+TEST_F(TestGPIOController, TestMotorsInit)
+{
+  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::VMOT_ON));
+  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::DRIVER_EN));
+}
+
+TEST_F(TestGPIOController, TestWatchdog)
+{
+  auto edge_cnt = std::make_shared<int>(0);
+
+  gpio_controller_wrapper_->RegisterGPIOEventCallback([this, edge_cnt](const auto & state) mutable {
+    if (state.pin == panther_gpiod::GPIOPin::WATCHDOG) {
+      (*edge_cnt)++;
+    }
+  });
+
+  gpio_controller_wrapper_->WatchdogEnable();
+  EXPECT_TRUE(gpio_controller_wrapper_->IsWatchdogEnabled());
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  gpio_controller_wrapper_->WatchdogDisable();
+
+  EXPECT_FALSE(gpio_controller_wrapper_->IsWatchdogEnabled());
+  EXPECT_EQ(*edge_cnt, watchdog_edges_per_100ms_ + 1);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  EXPECT_EQ(*edge_cnt, watchdog_edges_per_100ms_ + 1);
+}
+
+TEST_F(TestGPIOController, TestPinsAvailability)
+{
+  for (const auto & info : gpio_config_info_storage_) {
+    EXPECT_TRUE(gpio_controller_wrapper_->IsPinAvailable(info.pin));
+  }
+}
+
+TEST_F(TestGPIOController, TestFanEnbale)
+{
+  gpio_controller_wrapper_->FanEnable(true);
+  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::FAN_SW));
+
+  gpio_controller_wrapper_->FanEnable(false);
+  EXPECT_FALSE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::FAN_SW));
+}
+
+TEST_F(TestGPIOController, TestAUXPowerEnbale)
+{
+  gpio_controller_wrapper_->AUXPowerEnable(true);
+  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::AUX_PW_EN));
+
+  gpio_controller_wrapper_->AUXPowerEnable(false);
+  EXPECT_FALSE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::AUX_PW_EN));
+}
+
+TEST_F(TestGPIOController, TestChargerEnable)
+{
+  gpio_controller_wrapper_->ChargerEnable(true);
+  EXPECT_FALSE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::CHRG_DISABLE));
+
+  gpio_controller_wrapper_->ChargerEnable(false);
+  EXPECT_TRUE(gpio_controller_wrapper_->IsPinActive(panther_gpiod::GPIOPin::CHRG_DISABLE));
+}
+
+TEST_F(TestGPIOController, TestQueryControlInterfaceIOStates)
+{
+  std::unordered_map<panther_gpiod::GPIOPin, bool> io_states =
+    gpio_controller_wrapper_->QueryControlInterfaceIOStates();
+
+  EXPECT_EQ(io_states.size(), 7);
```

Comment:
If having failure here makes no sens in further testing, we could use assertion as well.